### PR TITLE
ORION-1821 feat: updates to oas package

### PR DESCRIPTION
**Pull request by Shockolate**

### CODE REVIEW

File: packages/oas/utils/json-schema-to-typescript-parsers/jsonParser.ts [hunk 1]
```diff
@@ -38,27 +37,29 @@ export const JsonParser = {
 	 * @returns {Promise}
 	 */
 	async parse(file: FileInfo): Promise<unknown> {
+		// eslint-disable-line require-await
```
Comment (matgr1) [inline packages/oas/utils/json-schema-to-typescript-parsers/jsonParser.ts:40]: don't think we need this

File: packages/oas/utils/json-schema-to-typescript-parsers/jsonParser.ts [hunk 2]
```diff
@@ -38,27 +37,29 @@ export const JsonParser = {
...
 		let data = file.data;
 		if (Buffer.isBuffer(data)) {
 			data = data.toString();
 		}
 
-		let parsed: JSONSchema4 | undefined;
 		if (typeof data === 'string') {
 			if (data.trim().length === 0) {
-				parsed = undefined; // This mirrors the YAML behavior
+				return; // This mirrors the YAML behavior
```
Comment (matgr1) [inline packages/oas/utils/json-schema-to-typescript-parsers/jsonParser.ts:48]:
```suggestion
				parsed = undefined; // This mirrors the YAML behavior
```

File: packages/oas/utils/json-schema-to-typescript-parsers/jsonParser.ts [hunk 3]
```diff
@@ -38,27 +37,29 @@ export const JsonParser = {
...
 			} else {
-				parsed = JSON.parse(data);
+				try {
+					return JSON.parse(data);
```
Comment (matgr1) [inline packages/oas/utils/json-schema-to-typescript-parsers/jsonParser.ts:51]:
```suggestion
					parsed = JSON.parse(data);
```

File: packages/oas/utils/json-schema-to-typescript-parsers/jsonParser.ts [hunk 4]
```diff
@@ -38,27 +37,29 @@ export const JsonParser = {
...
-			parsed = data;
-		}
 
-		if (parsed) {
-			return applyDateFormat(parsed);
+			if (data) {
+				return applyDateFormat(data);
+			}
+			return data;
 		}
```
Comment (matgr1) [inline packages/oas/utils/json-schema-to-typescript-parsers/jsonParser.ts:63]:
```suggestion
			parsed = data;
		}
		
	        if (parsed) {
		        return applyDateFormat(parsed);
	        }
```

File: packages/oas/utils/json-schema-to-typescript-parsers/jsonParser.ts [hunk 5]
```diff
@@ -38,27 +37,29 @@ export const JsonParser = {
 	 * @returns {Promise}
 	 */
 	async parse(file: FileInfo): Promise<unknown> {
+		// eslint-disable-line require-await
 		let data = file.data;
 		if (Buffer.isBuffer(data)) {
 			data = data.toString();
 		}
 
-		let parsed: JSONSchema4 | undefined;
```
Comment (matgr1) [inline packages/oas/utils/json-schema-to-typescript-parsers/jsonParser.ts:46]: ...also bring this back (see other code suggestions), we need to `applyDateFormat` when it does JSON.parse, as well as for the case when it's already parsed

File: packages/oas/utils/json-schema-to-typescript-parsers/yamlParser.ts [hunk 6]
```diff
@@ -41,24 +38,25 @@ export const YamlParser = {
 	 * @returns {Promise}
 	 */
 	async parse(file: FileInfo): Promise<unknown> {
+		// eslint-disable-line require-await
```
Comment (matgr1) [inline packages/oas/utils/json-schema-to-typescript-parsers/yamlParser.ts:41]: don't think we need this

File: packages/oas/utils/json-schema-to-typescript-parsers/yamlParser.ts [hunk 7]
```diff
@@ -41,24 +38,25 @@ export const YamlParser = {
...
 			// data is already a JavaScript value (object, array, number, null, NaN, etc.)
-			parsed = data;
-		}
 
-		if (parsed) {
-			return applyDateFormat(parsed);
+			if (data) {
+				return applyDateFormat(data);
+			}
+			return data;
```
Comment (matgr1) [inline packages/oas/utils/json-schema-to-typescript-parsers/yamlParser.ts:59]: similar to comments in jsonParser above, the old structure always did `applyDateFormat`, this only does if it was already parsed

File: packages/oas/utils/json-schema-to-typescript-parsers/yamlParser.ts [hunk 8]
```diff
@@ -1,8 +1,8 @@
...
-import JsonSchemaRefParser from 'json-schema-ref-parser';
+import { FileInfo } from '@apidevtools/json-schema-ref-parser';
+import JsonSchemaRefParser from '@apidevtools/json-schema-ref-parser';
+import { SchemaObject } from 'ajv';
 
-import { JSONSchema4 } from './JSONSchema4';
-import { applyDateFormat } from './applyDateFormat';
+import { applyDateFormat } from '../applyDateFormat';
 
 // NOTE: this is the yaml parser copied from v6.1.0 of json-schema-ref-parser (and modified) since it is not exported
```
Comment (matgr1) [inline packages/oas/utils/json-schema-to-typescript-parsers/yamlParser.ts]:
```suggestion
// NOTE: this is the yaml parser copied from v9.0.9 of json-schema-ref-parser (and modified) since it is not exported
```
maybe also include this link somewhere in the comment? https://github.com/APIDevTools/json-schema-ref-parser/blob/v9.0.9/lib/parsers/yaml.js

File: packages/oas/utils/json-schema-to-typescript-parsers/yamlParser.ts [hunk 9]
```diff
@@ -1,11 +1,11 @@
...
+import { load } from 'js-yaml';
 
-import { JSONSchema4 } from './JSONSchema4';
-import { applyDateFormat } from './applyDateFormat';
+import { applyDateFormat } from '../applyDateFormat';
 
-// NOTE: this is the yaml parser copied from v6.1.0 of json-schema-ref-parser (and modified) since it is not exported
-//       and this appears to be the only way to hook into parsing refs in order to fixup date fields
+// NOTE: this is the json parser copied from v9.0.9 of json-schema-ref-parser (and modified) since it is not exported
+// https://github.com/APIDevTools/json-schema-ref-parser/blob/v9.0.9/lib/parsers/yaml.js
```
Comment (matgr1) [inline packages/oas/utils/json-schema-to-typescript-parsers/yamlParser.ts:8]:
```suggestion
and this appears to be the only way to hook into parsing refs in order to fixup date fields
// https://github.com/APIDevTools/json-schema-ref-parser/blob/v9.0.9/lib/parsers/yaml.js
```
I think this part of the comment is important (it helped me remember the purpose at least :slightly_smiling_face:)

File: packages/oas/utils/json-schema-to-typescript-parsers/yamlParser.ts [hunk 10]
```diff
@@ -46,19 +46,23 @@ export const YamlParser = {
...
 
-		if (parsed) {
-			return applyDateFormat(parsed);
-		}
+			if (parsed) {
+				return applyDateFormat(parsed);
+			}
 
-		return parsed;
+			return parsed;
```
Comment (matgr1) [inline packages/oas/utils/json-schema-to-typescript-parsers/yamlParser.ts:66]: wonky tabbing?

File: packages/oas/utils/json-schema-to-typescript-parsers/yamlParser.ts [hunk 11]
```diff
@@ -1,11 +1,11 @@
...
+import { SchemaObject } from 'ajv';
+import { load } from 'js-yaml';
 
-import { JSONSchema4 } from './JSONSchema4';
-import { applyDateFormat } from './applyDateFormat';
+import { applyDateFormat } from '../applyDateFormat';
 
-// NOTE: this is the yaml parser copied from v6.1.0 of json-schema-ref-parser (and modified) since it is not exported
-//       and this appears to be the only way to hook into parsing refs in order to fixup date fields
+// NOTE: this is the json parser copied from v9.0.9 of json-schema-ref-parser (and modified) since it is not exported
```
Comment (matgr1) [inline packages/oas/utils/json-schema-to-typescript-parsers/yamlParser.ts:7]:
```suggestion
// NOTE: this is the yaml parser copied from v9.0.9 of json-schema-ref-parser (and modified) since it is not exported
```

**Review by matgr1 — DISMISSED** [review]

File: packages/service/src/controllers/Template.controller.ts [hunk 12]
```diff
@@ -57,7 +57,7 @@ export class TemplateController {
 	): Promise<ObjectIdModel> {
 		try {
 			this.logger.info(
-				{ message: _.pick(message, 'id', 'timestamp', 'source', 'destinations') },
+				{ template: _.pick(message, 'id', 'timestamp') },
```
Comment (Shockolate) [inline packages/service/src/controllers/Template.controller.ts:60]: This had to change because `logger.level` takes an object that MUST NOT have `message` or `msg` properties.

File: packages/service/src/models/ObjectId.model.ts [hunk 13]
```diff
@@ -13,4 +13,6 @@ export class ObjectIdModel extends Model implements ObjectId {
 	constructor(data?: Partial<ObjectIdModel>) {
 		super(data);
 	}
+
+	[k: string]: unknown;
```
Comment (Shockolate) [inline packages/service/src/models/ObjectId.model.ts:17]: Satisfies Typescript complaints. Can be silenced be `Pick` -ing the interface

File: packages/oas/utils/validateExamplesAgainstSchemas.ts [hunk 14]
```diff
@@ -4,44 +4,53 @@ import * as fs from 'fs';
 import { join, basename } from 'path';
 import { promisify } from 'util';
 
-import Ajv from 'ajv';
+import { SchemaObject } from 'ajv';
+import ajvFormats from 'ajv-formats';
+import Ajv2020 from 'ajv/dist/2020';
```
Comment (mgabeler-lee-6rs) [inline packages/oas/utils/validateExamplesAgainstSchemas.ts:9]: Is there a more "supported" way to import this, instead of diving into the `dist` directory? That kind of thing is very often ... brittle

File: packages/service/src/controllers/Template.controller.ts (hunk 12 above)
Comment (mgabeler-lee-6rs) [reply packages/service/src/controllers/Template.controller.ts:60]: was there a specific reason for removing `source` and `destination`? are they not helpful / too big?

File: packages/service/src/models/ObjectId.model.ts (hunk 13 above)
Comment (mgabeler-lee-6rs) [reply packages/service/src/models/ObjectId.model.ts:17]: The standard way we address this in other libs/repos is to change the `implements` to use `KnownFieldsOnly<T>`

File: packages/service/src/models/TemplateMessage.model.ts [hunk 15]
```diff
@@ -3,7 +3,7 @@ import { TemplateMessage } from '@sixriver/template-oas';
 
 // NOTE: the message is a `Model`, but _NOT_ an `Entity` -- it is not persisted
 @model()
-export class TemplateMessageModel extends Model implements TemplateMessage {
+export class TemplateMessageModel extends Model implements Pick<TemplateMessage, 'id' | 'data'> {
```
Comment (mgabeler-lee-6rs) [inline packages/service/src/models/TemplateMessage.model.ts:6]: `KnownKeysOnly<T>` again

File: .vscode/launch.json [hunk 16]
```diff
@@ -87,13 +87,16 @@
...
 			"outputCapture": "std",
 			"program": "${workspaceFolder}/packages/oas/dist/utils/convertModelsToSchemas.js",
 			"cwd": "${workspaceFolder}/packages/oas",
-			"skipFiles": ["<node_internals>/**"],
-			"preLaunchTask": "oas build utils"
+			"skipFiles": ["<node_internals>/**", "**/tslib/**/*.js"],
+			"preLaunchTask": "oas build utils",
+			"sourceMaps": true,
+			"stopOnEntry": false,
+			"resolveSourceMapLocations": ["${workspaceFolder}/**", "!**/node_modules/**"]
```
[the remark this answers is not in the document]
Comment (matgr1) [reply .vscode/launch.json:99]: good point, it won't break stepping into the libraries, but it would mean you'd end up in the `.js` instead... fyi this was a workaround for libraries that include source maps, but do not include source (in this case, vs code basically says "source code not found, but I'm not going to fall back to the .js, too bad")... I think we could probably limit this to the offending libs instead

File: .vscode/launch.json [hunk 17]
```diff
@@ -87,13 +87,16 @@
 		},
 		{
 			"name": "Run oas convertModelsToSchemas",
-			"type": "node",
+			"type": "pwa-node",
```
[the remark this answers is not in the document]
Comment (matgr1) [reply .vscode/launch.json:90]: this is the new default js debugger in vscode (https://github.com/microsoft/vscode-js-debug/)... it's required for the `resolveSourceMapLocations` setting below

File: packages/oas/utils/validateExamplesAgainstSchemas.ts (hunk 14 above)
Comment (matgr1) [reply packages/oas/utils/validateExamplesAgainstSchemas.ts:9]: was wondering the same thing, but this is from their example https://ajv.js.org/json-schema.html#draft-2020-12

File: .vscode/launch.json (hunk 17 above)
[the remark this answers is not in the document]
Comment (mgabeler-lee-6rs) [reply .vscode/launch.json:90]: I thought that was all ancient history now though, modern vscode releases I didn't think _have_ the old debugger any more? Trying to search info in this is hard, as most of the findings are from when this was definitely in preview/beta.

File: packages/oas/utils/validateExamplesAgainstSchemas.ts (hunk 14 above)
Comment (mgabeler-lee-6rs) [reply packages/oas/utils/validateExamplesAgainstSchemas.ts:9]: yeah, and some UTSL indicates that indeed the "main" doesn't require this file anywhere to re-export it, ok

File: .vscode/launch.json (hunk 17 above)
[the remark this answers is not in the document]
Comment (matgr1) [reply .vscode/launch.json:90]: not sure, I'm having trouble finding anything conclusive as well... but just did a quick test - adding a new launch config for `Node.js: Launch Program` (via the `Add Configuration...` button) does default to `pwa-node` using the latest version of vs code at least 🤷

[the remark this answers is not in the document]
Comment (mgabeler-lee-6rs) [reply .vscode/launch.json:90]: yeah, looking at the CI pipeline for vscode-js-debug, it seems the "--drop-in" mode is still not the default, so ... not sure what to make of it. I'm not clear on how that extension gets baked into the vscode core, so ... :shrug: this is probably fine as-is

[the remark this answers is not in the document]
Comment (matgr1) [reply .vscode/launch.json:90]: also, testing a mostly empty node js project (just a package.json and index.js), clicking the `create a launch.json file` from the debug tab will make it `node` if you select `Node.js (legacy)` and `pwa-node` if you select `Node.js`

File: packages/service/src/controllers/Template.controller.ts (hunk 12 above)
Comment (Shockolate) [reply packages/service/src/controllers/Template.controller.ts:60]: I don't think they were helpful, TemplateMessageModel does not have `source` nor `destinations`

File: packages/oas/utils/convertModelsToSchemas.ts [hunk 18]
```diff
@@ -51,28 +52,31 @@ function magicRefs(obj: StringKeyed) {
...
 
 	await Promise.all(
-		Object.keys(schemas).map(async (schemaName) => {
-			const schema = toJsonSchema(schemas[schemaName]);
-			const file = join(schemasPath, `${schemaName}.json`);
-			const content = JSON.stringify(schema, null, 2);
-			await write(schemaName, file, content);
-		}),
+		Object.keys(schemas)
+			.filter((schemaName) => /.+Message/g.test(schemaName))
```
[the remark this answers is not in the document]
Comment (Shockolate) [reply packages/oas/utils/convertModelsToSchemas.ts:66]: This filter will be removed

File: packages/oas/package.json [hunk 19]
```diff
@@ -27,30 +27,32 @@
...
+    "lint:spectral": "spectral lint template-openapi.yaml",
     "test:examples": "node dist/utils/validateExamplesAgainstSchemas.js",
     "unit": "true",
     "coverage": "true",
     "start": "false",
     "release": "true"
   },
   "dependencies": {
-    "@sixriver/typescript-support": "6.3.5"
+    "@sixriver/typescript-support": "9.0.0-HackDaysOAS.2"
```
Comment (Shockolate) [inline packages/oas/package.json:42]: TCA TODO - UPDATE ME

**Review by mgabeler-lee-6rs — COMMENTED** [review]
LGTM other than the already noted package.json updoots needed

**Review by mgabeler-lee-6rs — APPROVED** [review]
